### PR TITLE
#351, removed projects already added from list of repos in NewProject…

### PR DESCRIPTION
**Pull request by rkzel**

### CODE REVIEW

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js [hunk 1]
```diff
@@ -128,7 +132,11 @@ class Repo extends React.Component {
         </div>
       )
 
-    const repositories = repos.length > 0 ? repoList : <Loading />
+    // if there are repos to add - list them. otherwise either all repos
+    // are already added or we're still waiting for the list from external
+    const repositories = repos.length > 0 ? repoList :
```
Comment (Quazia) [inline apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Can we maybe just make this into a function? The chained ternary is sort of hard to read. I'm fine keeping it as is if you think it's better this way but to me it seems like a little too much logic to chain.

Comment (rkzel) [reply apps/projects/app/components/Panel/NewProject/RepoSelector.js]: You are right, nested ternaries are nasty. Improved (and saved myself from a lot of future confusion).

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js [hunk 2]
```diff
@@ -85,50 +91,54 @@ class Repo extends React.Component {
...
+        There are no repositories matching{' '}
+        <Text weight="bold">{this.state.filter}</Text>
+      </Text.Block>
+      <ClearSearch onClick={this.handleClearSearch}>
+        Clear Search
+      </ClearSearch>
+    </div>
+  )
+
+  stillLoading = () => !(this.props.data && this.props.data.viewer && this.props.data.viewer.repositories.edges.length > 0)
```
Comment (Quazia) [inline apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Can you try and reduce the line length here? This is still hard to read...

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js [hunk 3]
```diff
@@ -85,50 +91,54 @@ class Repo extends React.Component {
...
-          <Text.Block>
-            There are no repositories matching{' '}
-            <Text weight="bold">{this.state.filter}</Text>
-          </Text.Block>
-          <ClearSearch onClick={this.handleClearSearch}>
-            Clear Search
-          </ClearSearch>
-        </div>
-      )
+    // if there are repos to add - list them. otherwise either all repos
```
Comment (Quazia) [inline apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Much better for this part IMO

File: apps/projects/app/components/Panel/NewProject/NewProject.js [hunk 4]
```diff
@@ -27,7 +29,7 @@ class NewProject extends React.Component {
       : this.props.onGithubSignIn
 
     if (this.state.started)
-      return <RepoSelector onCreateProject={this.props.onCreateProject} />
+      return <RepoSelector onCreateProject={this.props.onCreateProject} reposAlreadyAdded={this.props.reposAlreadyAdded}/>
```
Comment (ottodevs) [inline apps/projects/app/components/Panel/NewProject/NewProject.js]: It seems you forgot an space character before closing tag

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js [hunk 5]
```diff
@@ -22,20 +22,26 @@ class Repo extends React.Component {
     repoSelected: -1,
   }
 
+  filterAlreadyAdded = repos => {
+    if (this.props.reposAlreadyAdded === undefined) return repos
+    const reposAlreadyAddedIds = this.props.reposAlreadyAdded.map(repo => repo.data._repo)
```
Comment (ottodevs) [inline apps/projects/app/components/Panel/NewProject/RepoSelector.js]: If we just need `repo.data._repo` fields for that, maybe we should just pass an array of this field as props instead of the whole repos data.
This way we should just need to use `Array.filter()` function to filter out repos with the same id as `reposAlreadyAddedIds`

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js [hunk 6]
```diff
@@ -22,20 +22,26 @@ class Repo extends React.Component {
...
+    return repos.filter(repo => reposAlreadyAddedIds.indexOf(repo.node.id) === -1)
```
Comment (ottodevs) [inline apps/projects/app/components/Panel/NewProject/RepoSelector.js]: IMHO `Array.some()` or `Array.includes()` seem more readable for these cases instead of `Array.indexOf()` and then compare index

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js [hunk 7]
```diff
@@ -83,50 +89,54 @@ class Repo extends React.Component {
...
+        There are no repositories matching{' '}
+        <Text weight="bold">{this.state.filter}</Text>
+      </Text.Block>
+      <ClearSearch onClick={this.handleClearSearch}>
+        Clear Search
+      </ClearSearch>
+    </div>
+  )
+
+  stillLoading = () => !(this.props.data && this.props.data.viewer && this.props.data.viewer.repositories.edges.length > 0)
```
Comment (ottodevs) [inline apps/projects/app/components/Panel/NewProject/RepoSelector.js]: [Apollo handles loading state on its own](https://www.apollographql.com/docs/react/essentials/queries.html#loading), so we should not use custom loading conditions at all

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js [hunk 8]
```diff
@@ -83,50 +89,54 @@ class Repo extends React.Component {
...
-        </div>
-      )
+    // if there are repos to add - list them. otherwise either all repos
+    // are already added or we're still waiting for the list from external
+    let repositoriesDisplay
 
-    const repositories = repos.length > 0 ? repoList : <Loading />
+    if (repos.length > 0)
+      repositoriesDisplay = this.repoList(visibleRepos, repoArray)
+    else if (this.stillLoading())
```
Comment (ottodevs) [inline apps/projects/app/components/Panel/NewProject/RepoSelector.js]: It maybe should be better to use[ Apollo loading state flow](https://www.apollographql.com/docs/react/essentials/queries.html#loading) here

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js [hunk 9]
```diff
@@ -265,6 +275,19 @@ const Loading = () => (
...
+  <div
+    style={{
+      display: 'flex',
+      justifyContent: 'center',
+      alignItems: 'center',
+      height: '20%',
+      flexDirection: 'column'
+    }}
+  >
+    <Text>No more repos to add...</Text>
```
Comment (ottodevs) [inline apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Maybe just an opinion, but for the sake of consistency we should use same denomination like "repositories" at L108, or change that if we decide to use "repos"

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js [hunk 10]
```diff
@@ -285,8 +308,8 @@ query {
    repositories(
      first: 100,
      orderBy: {field: UPDATED_AT, direction: DESC}
-     ownerAffiliations: [OWNER, COLLABORATOR, ORGANIZATION_MEMBER],
-     affiliations: [OWNER, COLLABORATOR, ORGANIZATION_MEMBER]) {
+     ownerAffiliations: [OWNER, COLLABORATOR],
```
Comment (ottodevs) [inline apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Could you please elaborate on why is this change needed?

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js (hunk 2 above)
Comment (rkzel) [reply apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Improved, although it still looks not very smart. There must be a better way of chaining conditions like that: if X and if X.Y and if X.Y.Z...

File: apps/projects/app/components/Panel/NewProject/NewProject.js (hunk 4 above)
Comment (rkzel) [reply apps/projects/app/components/Panel/NewProject/NewProject.js]: Duly noted :)

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js (hunk 9 above)
Comment (rkzel) [reply apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Repositories. Changed this line, consistency is important (and looks better).

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js (hunk 10 above)
Comment (rkzel) [reply apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Oh sorry, forgot to change it back after testing.

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js (hunk 5 above)
Comment (rkzel) [reply apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Done.

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js (hunk 6 above)
Comment (rkzel) [reply apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Improved.

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js (hunk 7 above)
Comment (rkzel) [reply apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Done.

File: apps/projects/app/components/Panel/NewProject/RepoSelector.js (hunk 8 above)
Comment (rkzel) [reply apps/projects/app/components/Panel/NewProject/RepoSelector.js]: Done.
